### ISSUE TITLE
Problems restoring stateful sets using velero plugin on vSphere with Tanzu

### ISSUE THREAD

**Issue by kurthv** [vmware-tanzu/velero-plugin-for-vsphere#305]
**Describe the bug**

I followed the instructions from Cormac Hogan (https://cormachogan.com/2021/03/03/vsphere-with-tanzu-stateful-application-backup-restore-using-velero-vsphere-operator/) to backup and restore a stateful application.
But the restore stucks and never gets finished.

From what I can see in vCenter, the PVC is showing up in the namespace folder, but is never created in container volumes.

**To Reproduce**

see the attached Readme, YAML and Log files

[backup-and-restore-cassandra.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6095249/backup-and-restore-cassandra.zip)

**Comment by lintongj** [comment]
@kurthv Thanks for reporting this issue.

Would you please the following information?

- Download and Clonefromsnapshot CRs. Please follow [this page](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/vanilla.md#restore) for detail if necessary.
- Velero Restore CR. e.g. `kubectl -n <velero namespace> get restore <restore CR name> -o yaml`
- Velero log about the restore. `kubectl -n <velero namespace> logs deploy/velero`

**Comment by kurthv** [comment]
Please see the attached archive 'logs.zip'

[logs.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6124595/logs.zip)

**Comment by lintongj** [comment]
@kurthv 
Thanks for sharing log bundle above. 

Initially, I thought it might be velero server that didn't notify vsphere plugin in the restore operation. Based on the log you provided, it is not the case. Somehow, vsphere plugin was not able to move restore forward. To investigate more, we need more information.

  - Backup Driver log. `kubectl -n <velero namespace> logs deploy/backup-driver`
  - Data Manager log(s). Based on your screenshot in the first thread, it seems to be a `vSphere with Tanzu` cluster. So, please perform checks by following the `Checking Velero Data Manager operations section` from [Cormac's blog](https://cormachogan.com/2021/03/03/vsphere-with-tanzu-stateful-application-backup-restore-using-velero-vsphere-operator/). After the checking is completed, collect data manager log via `docker logs velero-datamgr`.
  - VDDK log for each Data Manager. Login the  Data Manager VM as above. And run `docker cp <container id>:/tmp/vmware-root/. <any path on Data Manager VM>`
  - Plugin Upload and Download CRs. `kubectl -n <velero namespace> get <plugin crd name> -o yaml`
  - Plugin Snapshot and CloneFromSnapshot CRs. `kubectl -n <app namespace> get <plugin crd name> -o yaml`
 
One more tip here, please make sure you don't run the restore before the upload of backup is completed. Here is a [reference](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/vanilla.md#backup-vsphere-cns-block-volumes). Otherwise, restore will not work as expected.

**Comment by kurthv** [comment]
I will provide you with the requested information no later than the beginning of next week.

**Comment by kurthv** [comment]
Please see the attached log bundle.

Two notes:
1) The docker container on the Velero Data Manager didn't have a directory /tmp/vmware-root (but instead /tmp/vmware-nobody)
2) I wasn't sure, what I should use for <plugin crd name> in the last two KUBECTL commands? 
Therefore I used 'uploads' , 'downloads' and 'snapshots'.


[logs.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6136664/logs.zip)

**Comment by lintongj** [comment]
Thanks for sharing the log. In the zip file, logs.zip, above, your most recent backup, `cassandra-backup`, with 3 PVCs/PVs works as expected. However, from the downloads.txt, it was predated compared to your backup and uploads. Looks like you didn't try to restore your backup after backup is completed, according to the data manager log, `velero-datamgr.txt`. 

BTW, below is reason why the predated download fail.

```
"Failed to download snapshot, ivd:1cb8a5b7-b1f2-4f91-8ef2-413d93753534:eba12dd2-6a8c-4cf5-9bf4-182aad5db235,
      from durable object storage. GetObject failed for bucket velero, key plugins/vsphere-astrolabe-repo/ivd/peinfo/ivd:1cb8a5b7-b1f2-4f91-8ef2-413d93753534:eba12dd2-6a8c-4cf5-9bf4-182aad5db235:
      NoSuchKey: The specified key does not exist.\n\tstatus code: 404, request id:
      166B5A65644F3494, host id: "
```

The root cause is, the restore operation was before all the uploads were completed.

Please try the restore again. It should works as well. If not, please share log as the requests from both https://github.com/vmware-tanzu/velero-plugin-for-vsphere/issues/305#issuecomment-793002698 and https://github.com/vmware-tanzu/velero-plugin-for-vsphere/issues/305#issuecomment-797025430.

**Comment by kurthv** [comment]
Unfortunately this didn't work either :(
Attached you find a new log bundle.

[logs_2021-03-16.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6148812/logs_2021-03-16.zip)

**Comment by lintongj** [comment]
```
time="2021-03-16T11:19:23Z" level=error msg="Error creating volume from metadata" error="failed to create PVC cassandra/cassandra-data-cassandra-0 in the API server: namespaces \"cassandra\" not found" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:769"
...
time="2021-03-16T11:32:55Z" level=error msg="Failed at calling SnapshotManager cloneFromSnapshot with peId pvc:cassandra/cassandra-data-cassandra-0" controller=BackupDriverController error="Error creating volume from metadata: failed to create PVC cassandra/cassandra-data-cassandra-0 in the API server: persistentvolumeclaims \"cassandra-data-cassandra-0\" is forbidden: insufficient quota to consume: tanzu-storage-policy.storageclass.storage.k8s.io/requests.storage" error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr/snapshot_manager.go:770" error.function="github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/snapshotmgr.(*SnapshotManager).CreateVolumeFromSnapshotWithMetadata" logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backupdriver/backup_driver_controller.go:234"
```

Found the log messages above in backup-driver log at `backup-driver.txt`. Please confirm that the `cassandra` namespace is not available in the cluster. If that is the case, I think I get the root cause. vSphere with tanzu cluster (aka, supervisor cluster) is special. Before running `velero restore`, the target namespace is expected to be created (via vSphere UI or vSphere DCLI, NOT from kubernetes cluster) and the storage policy is also expected to be configured within the target namespace (otherwise, the `insufficient quota to consume` error will be reported). 

Please have a try. And we will update our documentation.

**Comment by kurthv** [comment]
Hi Lintong Jiang,

unfortunately this was not the root cause.
The namespace & storage policy does exist.
In fact I only deleted the cassandra stateful set and the pvc after backup and before restore.

But I found some some other hints:
`  status:
    message: 'Failed to download snapshot, ivd:5530f7cf-887f-4b07-b820-6ea6bb482b12:6eefb5b7-78e4-4c8b-9f60-dfd0da8ea0a1,
      from durable object storage. Data copy failed from PE ivd:fc47adf9-6ff9-4a25-a999-a255eef79a1b,
      to PE ivd:5530f7cf-887f-4b07-b820-6ea6bb482b12:6eefb5b7-78e4-4c8b-9f60-dfd0da8ea0a1:
      Failed in CopyBuffer, bytes written = 1048576: Write to virtual disk file failed.
      The error code is 14009.'
    nextRetryTimestamp: "2021-03-17T11:34:26Z"
    phase: InProgress
`

At the same time I see in vixDiskLib-1:
`2021-03-17T11:19:11.007Z| host-36| I125: VixDiskLib: VixDiskLib_OpenEx: Open a disk.
2021-03-17T11:19:11.142Z| host-36| I125: NBD_ClientOpen: attempting to create connection to ha-nfc://[vsanDatastore] f7ee4960-7209-2b88-9844-005056963e0c/807da2608f08408cb3ad79be57de3ad1.vmdk@{nfcip:172.16.13.24,hostname:esx0d.nested.lab}:902
2021-03-17T11:19:11.276Z| host-36| I125: Setting NFC log level to 1
2021-03-17T11:19:11.276Z| host-36| I125: Setting NFC log level to 1
2021-03-17T11:19:11.277Z| host-36| I125: NFC Async IO session is established with log level 1.
2021-03-17T11:19:11.277Z| host-36| I125: Opening file [vsanDatastore] f7ee4960-7209-2b88-9844-005056963e0c/807da2608f08408cb3ad79be57de3ad1.vmdk (ha-nfc://[vsanDatastore] f7ee4960-7209-2b88-9844-005056963e0c/807da2608f08408cb3ad79be57de3ad1.vmdk@{nfcip:172.16.13.24,hostname:esx0d.nested.lab}:902)
2021-03-17T11:19:11.285Z| host-36| I125: DISKLIB-LINK  : Opened 'ha-nfc://[vsanDatastore] f7ee4960-7209-2b88-9844-005056963e0c/807da2608f08408cb3ad79be57de3ad1.vmdk@{nfcip:172.16.13.24,hostname:esx0d.nested.lab}:902' (0x800a): custom, 204800 sectors / 100 MB.
2021-03-17T11:19:11.289Z| host-36| I125: DISKLIB-LIB   : Opened "ha-nfc://[vsanDatastore] f7ee4960-7209-2b88-9844-005056963e0c/807da2608f08408cb3ad79be57de3ad1.vmdk@{nfcip:172.16.13.24,hostname:esx0d.nested.lab}:902" (flags 0x800a, type custom).
2021-03-17T11:19:11.292Z| host-36| I125: VixDiskLib: VixDiskLib_GetInfo: Retrieve disk info.
2021-03-17T11:19:11.306Z| host-36| I125: VixDiskLib: VixDiskLib_FreeInfo: Clean up VixDiskLib.
2021-03-17T11:24:17.677Z| host-43| W115: [NFC ERROR]NfcNetTcpSetError: Connection reset by peer
2021-03-17T11:24:17.677Z| host-43| W115: [NFC ERROR]NfcNetTcpWrite: bWritten: -1. Errno: 104.
2021-03-17T11:24:17.678Z| host-43| W115: [NFC ERROR]NfcAioSendData: Failed to send data: NFC_NETWORK_ERROR
2021-03-17T11:24:17.678Z| host-43| W115: [NFC ERROR]NfcAioLogFatalSessionErrorLocked: A fatal session error occurred. The error was: 'NFC_NETWORK_ERROR' (3)
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcNetTcpSetError: Unexpected EOF
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcNetTcpRead: EOF should not be reached
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcNet_Recv: requested 16, recevied only 0 bytes
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcAioRecvData: Failed to receive data: NFC_NETWORK_ERROR
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcAioLogFatalSessionErrorLocked: A fatal session error occurred. The error was: 'NFC_NETWORK_ERROR' (3)
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcAioGetMessage: Recv msg failed: NFC_NETWORK_ERROR
2021-03-17T11:24:17.678Z| host-44| W115: [NFC ERROR]NfcAioClientReceiveLoop: Failed to receive an AIO message: NFC_NETWORK_ERROR
2021-03-17T11:24:17.678Z| host-43| W115: [NFC ERROR]NfcAioSendMessage: Cln send failed: NFC_NETWORK_ERROR
2021-03-17T11:24:17.678Z| host-43| W115: [NFC ERROR]NfcAioSndIoMsg: Failed to send an IO message: NFC_NETWORK_ERROR
2021-03-17T11:24:17.678Z| host-44| I125: DISKLIB-LIB   : RWv failed ioId: #2 (2338) (34) .
2021-03-17T11:24:17.679Z| host-22| E110: VixDiskLib: Detected DiskLib error 2338 (NBD_ERR_NETWORK_CONNECT).
2021-03-17T11:24:17.679Z| host-22| E110: VixDiskLib: VixDiskLib_Write: Write 2048 sectors to disk at 0 failed. Error 14009 (The server refused connection) (DiskLib error 2338: NBD_ERR_NETWORK_CONNECT) at 6296.
2021-03-17T11:24:17.681Z| host-22| I125: VixDiskLib: VixDiskLib_Close: Close disk.
`
See also the attached log bundle.

[logs_2021-03-17.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6156090/logs_2021-03-17.zip)

**Comment by lintongj** [comment]
It should not be a network issue about port 902 is not opened. I still suspect the target namespace and the storage policy in the  namespace is not configured as expected.

Before trying another restore operation to the target namespace, would you please try to provision a new PVC in the target namespace and see it works?

**Comment by kurthv** [comment]
Provisioning PVCs via KUBECTL is not an issue at all.
In fact, when I delete the downloads with `kubectl -n velero-ns delete downloads --all` the restore operation proceeds and new PVCs are created.

I found quite a number of similiar issues, when searching for "VMware NFC" and "asynchronous IO" whith "error 14009":

- https://kb.vmware.com/s/article/71046
- https://github.com/vmware-tanzu/velero-plugin-for-vsphere/issues/145
- https://docs.druva.com/Knowledge_Base/Phoenix/Troubleshooting/VMware_backups_fail_with_NFC_errors

But I am using the latest version of ESXi -> vSphere 7 Update 2 (Build 17630552) with vSAN Datastore

**Comment by lintongj** [comment]
> Provisioning PVCs via KUBECTL is not an issue at all.
> In fact, when I delete the downloads with `kubectl -n velero-ns delete downloads --all` the restore operation proceeds and new PVCs are created.
> 

So, do you mean the restore finally works? That's cool. Please try backup and restore for another stateful application in supervisor cluster and make sure it works as expected. 

> I found quite a number of similiar issues, when searching for "VMware NFC" and "asynchronous IO" whith "error 14009":
> 
> * https://kb.vmware.com/s/article/71046
> * #145
> * https://docs.druva.com/Knowledge_Base/Phoenix/Troubleshooting/VMware_backups_fail_with_NFC_errors
> 
> But I am using the latest version of ESXi -> vSphere 7 Update 2 (Build 17630552) with vSAN Datastore

As long as you followed our doc to [setup Backup Network](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/main/docs/supervisor-datamgr.md#setting-up-backup-network) and open expected port, 902, on ESXs. The 14009 error should not be related with network configuration. I will ask VDDK folks to debug this issue though.

**Comment by kurthv** [comment]
No, restore doesn‘t work!

When I execute `Kubectl delete downloads —all`the download is stopped and NEW persistent volumes are created.
Data is lost, because the original PVs are not restored.

Please have a look at the log bundle from 03/17/2021.

**Comment by lintongj** [comment]
> No, restore doesn‘t work!
> 
> When I execute `Kubectl delete downloads —all`the download is stopped and NEW persistent volumes are created.
> Data is lost, because the original PVs are not restored.
> 
> Please have a look at the log bundle from 03/17/2021.

I see. That's pretty weird issue we've never seen before. Basically, you can upload snapshot data to object store which means  your network connection works as expected. However, the download of snapshot from object store just fails. 

Would you please also help collect vpxa logs from ESX host, `esx0d.nested.lab`? That might be helpful for VDDK folks to debug this issue. I will update once get any feedback from VDDK folks.

**Comment by kurthv** [comment]
Attached you find the logs from today (including vpxa.log).

[logs_2021-03-19.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6173601/logs_2021-03-19.zip)

**Comment by lintongj** [comment]
> Attached you find the logs from today (including vpxa.log).
> 
> [logs_2021-03-19.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6173601/logs_2021-03-19.zip)


>Would you please also help collect vpxa and hostd logs from ESX host, esx0d.nested.lab?

Would you please collect **hostd** log as well? Thanks

**Comment by kurthv** [comment]
Here are the logs from yesterday - including hostd.log:

[logs_2021-03-20.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6175225/logs_2021-03-20.zip)

**Comment by lintongj** [comment]
Hi @kurthv, I got feedback from VDDK folks. Simply speaking, there might be some network configuration issue in your environment.

Based on the observation so far,
- Upload works OK, which means outgoing traffic via port 902 on ESX works as expected.
- Download fails consistently, which means incoming traffic via port 902 on ESX doesn't work well.

Would you please double check if there is specific configuration in your network that prevent incoming traffic in ESX?

Also, would you please share the detail about what did you do to setup the Backup Network in your supervisor cluster?

**Comment by kurthv** [comment]
I did exactly follow the instructions by Cormac Hogan to setup the backup network.
And when I look at hostd.log it seems to me as an issue with writing to the PV vmdk.

**Comment by lintongj** [comment]
```
And when I look at hostd.log it seems to me as an issue with writing to the PV vmdk.
```
Yes, there is an issue with writing to the PV vmdk, but it is very likely due to network issue. `writing to the PV vmdk` has been extensively tested and we've never seen a similar one before.

**Comment by kurthv** [comment]
I will now stop further investigations :(
It seems, that this is not yet "production ready" and needs some more development and intense testing.

**Comment by kurthv** [comment]
BTW 
Is there a tool available to test these kind of write I/O over the network?

**Comment by lintongj** [comment]
> I will now stop further investigations :(
> It seems, that this is not yet "production ready" and needs some more development and intense testing.

velero-plugin-for-vsphere has been intensively tested and is production ready, but we cannot guarantee the coverage of all users' setup. 

The issue you reported is valuable for us to learn and improve. So far, it doesn't seem to be a bug in the plugin, but more likely one in network configuration. I have filed a bug internally to track this issue and engaged with vSphere team to investigate this issue furthermore. We may request ESXi log bundles from you if necessary. For the moment, would you please confirm the following information.

- VC & ESX versions. Just to confirm if VCenter and ESXs are vSphere 7 Update 2.
- Port 902 is opened on every ESXi for both incoming and outgoing traffic. (You can check it via vSphere UI. For each ESXi host, navigate to the **Configure** tab and select **Firewall** under the **System** section. There are INCOMING and OUTGOING tabs. Double if **NFC** service along with port **902** is configured for both directions. Please refer to the screenshot below.)

![Screen Shot 2021-03-23 at 11 07 39 AM](https://user-images.githubusercontent.com/55512168/112196457-4b6bad00-8bc8-11eb-82f6-4ba2c2b13ac0.png)


> BTW
> Is there a tool available to test these kind of write I/O over the network?

[VDDK](https://code.vmware.com/web/sdk/7.0/vddk) has a command line tool for VMDK read/write. But, there is some learning curve to use it. I am also thinking of building a handy tool for such test. Will keep you posted.

**Comment by kurthv** [comment]
> VC & ESX versions. Just to confirm if VCenter and ESXs are vSphere 7 Update 2.

YES

> Port 902 is opened on every ESXi for both incoming and outgoing traffic. (You can check it via vSphere UI. For each ESXi host, navigate to the Configure tab and select Firewall under the System section. There are INCOMING and OUTGOING tabs. Double if NFC service along with port 902 is configured for both directions. Please refer to the screenshot below.)

YES

![image](https://user-images.githubusercontent.com/72090771/112281585-80d6d000-8c86-11eb-9a03-e81c1c5d80cb.png)

![image](https://user-images.githubusercontent.com/72090771/112281754-b24f9b80-8c86-11eb-9379-0d46b2be357f.png)

**Comment by kurthv** [comment]
I have one final idea:
![image](https://user-images.githubusercontent.com/72090771/112323276-ed1af900-8cb1-11eb-9f07-64b2215b29a2.png)

Is it possible that the docker container is running in the wrong security context (nobody)?

**Comment by lintongj** [comment]
> I have one final idea:
> ![image](https://user-images.githubusercontent.com/72090771/112323276-ed1af900-8cb1-11eb-9f07-64b2215b29a2.png)
> 
> Is it possible that the docker container is running in the wrong security context (nobody)?

That's expected. Otherwise, upload during the backup should have not worked.

**Comment by lintongj** [comment]
Hi @kurthv, would you please help reproduce this issue and collect following information, so that vSphere NFC team can help debug this issue better?

Before you reproduce it, please set ESXi **hostd** log level to **trivia**. Please refer to https://kb.vmware.com/s/article/1004795 for detail. Also, run packet capture tool, like tcpdump, on data manager VM to capture NFC traffic by restricting port to 902. 

Then, run velero restore in your supervisor cluster. When you see download CR is failed, please collect the following information.

Below is the information to collect.
- collect ESXi support bundle. you can do that by running `vm-support` after ssh into the ESXi host where the restored PodVM is on.
- collect packet capture result from the client side, i.e., data manager VM.

Thanks!

**Comment by kurthv** [comment]
Here the results of the TCPDUMP:
[logs_2021-03-31.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6237744/logs_2021-03-31.zip)

The Support Bundle has a size of 393MB.
Therefore I am not able to upload the tarball.

**Comment by kurthv** [comment]
172.16.13.22 is the IP of the VMKernel where NFC is enabled. It‘s in the same subnet as the Velero Data Manager VM (172.16.13.50).

The TCPDUMP was created on the Velero Datamanger VM with the following command:

`tcpdump -i eth0 -w dump.bin ‘Port 902‘`

**Comment by kurthv** [comment]
No more updates on this issue?

**Comment by lintongj** [comment]
NFC team doesn't have much clue based on the log provided earlier, and is asking for more logs. 

We don't want to go back and forth and ask for logs from you again and again. So, we'd like to develop a tool(hopefully early next week) to test VMDK I/O over the network before asking for more logs from you.

**Comment by kurthv** [comment]
For me it's pretty clear, that Velero Data Manager ignores the vmkernel interface that is tagged for NFC.
Instead it tries to use the MGMT vmkernel interface, but then traffic needs to pass a firewall which most likely causes issues.

**Comment by lintongj** [comment]
> For me it's pretty clear, that Velero Data Manager ignores the vmkernel interface that is tagged for NFC.
> Instead it tries to use the MGMT vmkernel interface, but then traffic needs to pass a firewall which most likely causes issues.

Right, that could be the culprit. I will further confirm with the team about why traffic is rerouted to MGMT network while the service, `vSphere Backup NFC`, is enabled on the backup network.

BTW, do you have some specific filewall rule on your MGMT network? Such as, only allow outgoing traffic while disallowing incoming traffic. Otherwise, I would be surprised about why upload works fine while download doesn't in your setup.

**Comment by kurthv** [comment]
There is an ANY-ANY-ALLOW rule between MGMT and Backup/NFC network on my pfSense firewall.
But I guess, that the asynchronous I/O is causing issues when routed through a firewall.

**Comment by lintongj** [comment]
>From what I can see, NFC traffic is not send to the nfcip:172.16.13.22
>but instead to the MGMT-vmk esx0b.nested.lab, 172.16.254.22:

I got it confirmed with VDDK team. In the current version of VDDK(7.0) used in velero-plugin-for-vsphere, there is a bug that the NFC traffic will be unexpectedly rerouted to MGMT network even though Backup network with `vSphere Backup NFC` service is configured. It is fixed in 7.0U1 in VDDK. It is also a bug in velero-plugin-for-vsphere as well and we need to pick up 7.0U1(or above) VDDK libraries and fix this issue in our next release.

> There is an ANY-ANY-ALLOW rule between MGMT and Backup/NFC network on my pfSense firewall.
> But I guess, that the asynchronous I/O is causing issues when routed through a firewall.

What asynchronous I/O do you mean? If you mean about the VDDK async I/O, actually our wrapper of VDDK doesn't use any async I/O APIs even though they are available in VDDK.

Also, what's the firewall rule between MGMT network and Velero Data Manager?  If NFC traffic(via port 902) from Data Manager is allowed, we should not see the failure of restore/download.

**Comment by lintongj** [comment]
>Also, what's the firewall rule between MGMT network and Velero Data Manager? If NFC traffic(via port 902) from Data Manager is allowed, we should not see the failure of restore/download.

@kurthv Would you please help double check your setup based on the ask above? So that we can confirm there is no any other issue apart from the known VDDK bug. Thanks.

**Comment by kurthv** [comment]
> There is an ANY-ANY-ALLOW rule between MGMT and Backup/NFC network on my pfSense firewall.
> But I guess, that the asynchronous I/O is causing issues when routed through a firewall.

Like I already wrote, there is no firewall rule blocking the traffic (ANY-to-ANY-ALLOW-for-all-ports).

**Comment by lintongj** [comment]
> As a workaround I don't use a dedicated backup network. Instead I use the management network also for backup traffic, and the Velero Datamanager is also located in the management network.

Does velero restore work well with the workaround?

**Comment by lintongj** [comment]
Closed it as [the release v1.2.0](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/releases/tag/v1.2.0) was out.
